Code/Framework/AzToolsFramework/AzToolsFramework/ViewportSelection/ViewportEditorModeTracker.cpp

```diff
@@ -0,0 +1,151 @@
+        else
+        {
+            return AZ::Failure(
+                AZStd::string::format(ViewportEditorModeLogWindow, false, "Cannot deactivate mode %u, mode is not recognized", modeIndex));
+        }
+    }
+
+    bool ViewportEditorModes::IsModeActive(ViewportEditorMode mode) const
+    {
+        return m_editorModes[static_cast<AZ::u32>(mode)];
+    }
+
+    void ViewportEditorModeTracker::RegisterInterface()
+    {
+        if (AZ::Interface<ViewportEditorModeTrackerInterface>::Get() == nullptr)
+        {
+            AZ::Interface<ViewportEditorModeTrackerInterface>::Register(this);
+        }
+    }
+
+    void ViewportEditorModeTracker::UnregisterInterface()
+    {
+        if (AZ::Interface<ViewportEditorModeTrackerInterface>::Get() != nullptr)
+        {
+            AZ::Interface<ViewportEditorModeTrackerInterface>::Unregister(this);
+        }
+    }
+
+    AZ::Outcome<void, AZStd::string> ViewportEditorModeTracker::RegisterMode(
+        const ViewportEditorModeInfo& viewportEditorModeInfo, ViewportEditorMode mode)
+    {
+        auto& editorModes = m_viewportEditorModesMap[viewportEditorModeInfo.m_id];
+        if (editorModes.IsModeActive(mode))
+        {
+            return AZ::Failure(AZStd::string::format(
+                "Duplicate call to RegisterMode for mode '%u' on id '%i'", static_cast<AZ::u32>(mode), viewportEditorModeInfo.m_id));
+        }
+        
+        if (const auto result = editorModes.SetModeActive(mode);
+            !result.IsSuccess())
+        {
+            return result;
+        }
+
+        ViewportEditorModeNotificationsBus::Event(
+            viewportEditorModeInfo.m_id, &ViewportEditorModeNotificationsBus::Events::OnEditorModeEnter, editorModes, mode);
+
+        return AZ::Success();
+    }
+
+    AZ::Outcome<void, AZStd::string> ViewportEditorModeTracker::UnregisterMode(
+        const ViewportEditorModeInfo& viewportEditorModeInfo, ViewportEditorMode mode)
+    {
+        ViewportEditorModes* editorModes = nullptr;
+        bool modeWasActive = true;
+        if (m_viewportEditorModesMap.count(viewportEditorModeInfo.m_id))
+        {
+            editorModes = &m_viewportEditorModesMap.at(viewportEditorModeInfo.m_id);
+            if (!editorModes->IsModeActive(mode))
+            {
+                return AZ::Failure(AZStd::string::format(
+                    "Duplicate call to UnregisterMode for mode '%u' on id '%i'", static_cast<AZ::u32>(mode), viewportEditorModeInfo.m_id));
+            }
+        }
+        else
+        {
+            modeWasActive = false;
+            editorModes = &m_viewportEditorModesMap[viewportEditorModeInfo.m_id];
+        }
+
+        if(const auto result = editorModes->SetModeInactive(mode);
+            !result.IsSuccess())
+        {
+            return result;
+        }
+
+        ViewportEditorModeNotificationsBus::Event(
+            viewportEditorModeInfo.m_id, &ViewportEditorModeNotificationsBus::Events::OnEditorModeExit, *editorModes, mode);
```

Comment:
update: If we're going to call this Register/Unregister (that this should be `OnEditorModeUnregistered` too - but I'm not clear what would actually care about a mode being registered/unregistered whereas entering/exiting means something different to me and is very useful (maybe this is the Active/Inactive thing now?)

Reply:
sure, see above 👍